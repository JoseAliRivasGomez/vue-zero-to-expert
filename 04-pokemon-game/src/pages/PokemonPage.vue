<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1>Quien es este Pokemon?</h1>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />

        <template v-if="showAnswer">
            <h2 class="fade-in">{{message}}</h2>
            <button @click="newGame">Nuevo Juego</button>
        </template>
        
    </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()
            const randInt = Math.floor(Math.random()*4)
            this.pokemon = this.pokemonArr[randInt]
        },
        checkAnswer(selectedId) {
            this.showPokemon = true
            this.showAnswer = true

            if(selectedId === this.pokemon.id){
                this.message = `Correcto, ${this.pokemon.name}`
            }else{
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame() {
            this.pokemonArr = []
            this.pokemon = null
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''
            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>