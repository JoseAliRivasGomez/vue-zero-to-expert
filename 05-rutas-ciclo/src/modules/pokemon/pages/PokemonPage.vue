<template>
    <h1>Pokemon #<span>{{id}}</span></h1>
    <div v-if="pokemon">
        <img :src="pokemon.sprites.front_default" :alt="pokemon.name">
    </div>
</template>

<script>
export default {
    props: {
        id: {
            type: Number,
            required: true
        }
    },
    data(){
        return {
            //id: null
            //id: this.$route.params.id
            pokemon:+ null
        }
    },
    created(){
        //console.log(this.$route)
        //const {id} = this.$route.params
        //console.log(id)
        //this.id = id
        this.getPokemon()
    },
    methods: {
        async getPokemon() {
            try {
                const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`).then(r => r.json())
                //console.log(pokemon);
                this.pokemon = pokemon
            } catch (error) {
                this.$router.push('/')
            }
        }
    },
    watch: {
        id(){
            this.getPokemon()
        }
    }
}
</script>