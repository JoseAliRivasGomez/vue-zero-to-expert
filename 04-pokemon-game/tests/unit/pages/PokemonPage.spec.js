import { mount, shallowMount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage";
import { pokemons } from "../mocks/pokemons.mock";


describe('PokemonPage', () => { 

    let wrapper
    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('Debe de hacer match con el snapshot', () => { 

        expect(wrapper.html()).toMatchSnapshot() //or toMatchInlineSnapshot

    })

    test('Debe de llamar el mixPokemonArray al montar', () => { 

        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
        
    })

    test('Debe de hacer match con el snapshot cuando cargan los pokemons', () => { 

        const wrapper = shallowMount(PokemonPage, { //el mount monta los componentes PokemonOptions y PokemonPicture, el shallowMount no
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
        
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => { 

        const wrapper = shallowMount(PokemonPage, { //el mount monta los componentes PokemonOptions y PokemonPicture, el shallowMount no
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()

        expect(picture.attributes('pokemonid')).toBe('1')
        expect(options.attributes('pokemons')).toBeTruthy()
        
    })

    test('Pruebas con checkAnswer', async () => { 

        const wrapper = shallowMount(PokemonPage, { //el mount monta los componentes PokemonOptions y PokemonPicture, el shallowMount no
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[1],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(2)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, ${pokemons[1].name}`)

        await wrapper.vm.checkAnswer(100)

        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[1].name}`)

    })

})