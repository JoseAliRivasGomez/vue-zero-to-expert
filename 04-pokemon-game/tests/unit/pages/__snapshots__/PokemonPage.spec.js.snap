// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>Quien es este Pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
